test/Microsoft.AspNet.Mvc.ModelBinding.Test/Metadata/ModelMetadataAttributesTest.cs

```diff
@@ -0,0 +1,158 @@
+        {
+            // Arrange
+            var modelType = typeof(DerivedViewModel);
+            var property = modelType.GetRuntimeProperties().FirstOrDefault(p => p.Name == "TestProperty");
+
+            // Act
+            var attributes = ModelAttributes.GetAttributesForProperty(modelType, property);
+
+            // Assert
+            Assert.Single(attributes.OfType<RequiredAttribute>());
+            Assert.Single(attributes.OfType<StringLengthAttribute>());
+            Assert.DoesNotContain(typeof(RangeAttribute), attributes);
+        }
+
+        [Fact]
+        public void GetAttributesForVirtualPropertyFromDerived_IncludesMetadataAttributes()
+        {
+            // Arrange
+            var modelType = typeof(DerivedViewModel);
+            var property = modelType.GetRuntimeProperties().FirstOrDefault(p => p.Name == "VirtualProperty");
+
+            // Act
+            var attributes = ModelAttributes.GetAttributesForProperty(modelType, property);
+
+            // Assert
+            Assert.Single(attributes.OfType<RequiredAttribute>());
+            Assert.Single(attributes.OfType<RangeAttribute>());
+        }
+
+        [Fact]
+        public void GetAttributesForType_IncludesMetadataAttributes()
+        {
+            // Arrange & Act
+            var attributes = ModelAttributes.GetAttributesForType(typeof(BaseViewModel));
+
+            // Assert
+            Assert.Single(attributes.OfType<ClassValidator>());
+        }
+
+        // Helper classes
+
+        [ClassValidator]
+        private class BaseModel
+        {
+            [StringLength(10)]
+            public string BaseProperty { get; set; }
+
+            [Range(10,100)]
+            public string TestProperty { get; set; }
+
+            [Required]
+            public virtual int VirtualProperty { get; set; }
+        }
+
+        private class DerivedModel : BaseModel
+        {
+            [Required]
+            public string DerivedProperty { get; set; }
+
+            [Required]
+            public new string TestProperty { get; set; }
+
+            [Range(10,100)]
+            public override int VirtualProperty { get; set; }
+            
+        }
+
+        [ModelMetadataType(typeof(BaseModel))]
+        private class BaseViewModel
+        {
+            [Range(0,10)]
+            public string TestProperty { get; set; }
+
+            [Required]
+            public string BaseProperty { get; set; }
+        }
+
+        [ModelMetadataType(typeof(DerivedModel))]
+        private class DerivedViewModel : BaseViewModel
+        {
+            [StringLength(2)]
+            public new string TestProperty { get; set; }
+
+            public int VirtualProperty { get; set; }
+
+        }
+
+        private class ClassValidator : ValidationAttribute
+        {
+            public override Boolean IsValid(Object value)
+            {
+                return true;
+            }
+        }
```

Comment:
would like to see some of the binder attributes as well, since that's affected by this change. (`[FromHeader]` and friends).